presto-hive/src/main/java/com/facebook/presto/hive/util/CloseableAsyncQueue.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.util;
+
+import com.google.common.collect.ImmutableList;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@ThreadSafe
+public class CloseableAsyncQueue<T>
+{
+    private static final CompletableFuture<?> COMPLETED_FUTURE_OF_NULL = CompletableFuture.completedFuture(null);
+
+    private final AsyncQueue<T> queue;
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+
+    private final CompletableFuture<List<T>> completedFutureOfEmptyList = CompletableFuture.completedFuture(ImmutableList.of());
+
+    public CloseableAsyncQueue(int softMaxQueueSize)
+    {
+        queue = new AsyncQueue<>(softMaxQueueSize);
+    }
+
+    public int size()
+    {
+        return queue.size();
+    }
+
+    /**
+     * After calling this function, {@link #isClosed()} will return true, and subsequent calls to
+     * {@link #offer(Object)} will be ignored. In addition, all futures returned previously will be
+     * immediately completed. Subsequent calls to {@link #offer(Object)} and {@link #getBatchAsync(int)}
+     * will always return a completed future. Ongoing calls to {@link #offer(Object)} may or may not be ignored.
+     */
+    public void close()
+    {
+        if (closed.compareAndSet(false, true)) {
+            // It must be guaranteed that no one will become blocked once execution hits this point.
+            queue.signalAll();
+        }
+    }
+
+    public boolean isClosed()
+    {
+        return closed.get();
+    }
+
+    public CompletableFuture<?> offer(T element)
+    {
+        if (closed.get()) {
+            return COMPLETED_FUTURE_OF_NULL;
+        }
+        return queue.offer(element);
```

Comment:
If `queue` is full, a thread T could be waiting to execute this line when `close()` is called. It would then miss the `signalAll()` call and receive a non-completed future. To fix this I think you need to reverse the order of the `if` and this `offer()` call (introducing a local variable for the return value of `offer()`)